runtime/src/stakes.rs

```diff
@@ -62,28 +62,44 @@ impl StakesCache {
         // but the owner changes, then this needs to evict the account from
         // the cache. see:
         // https://github.com/solana-labs/solana/pull/24200#discussion_r849935444
-        if solana_vote_program::check_id(account.owner()) {
-            let new_vote_account = if account.lamports() != 0
-                && VoteState::is_correct_size_and_initialized(account.data())
-            {
+        let owner = account.owner();
+        // Zero lambort accounts are not stored in accounts-db
+        // and so should be removed from cache as well.
+        if account.lamports() == 0 {
+            if solana_vote_program::check_id(owner) {
+                let mut stakes = self.0.write().unwrap();
+                stakes.remove_vote_account(pubkey);
+            } else if solana_stake_program::check_id(owner) {
```

Comment:
Before your change, `update_stake_delegation` would still store an initialized delegation even if it had zero lamports. However, the stake program always sets state to uninitialized when it drains a stake account to zero lamports. It also doesn't allow non-rent-exempt stake accounts to be created. So this looks safe to me without a feature gate.

Reply:
yup, that is consistent with my thinking.
Are there any places then that rely on this logic?
https://github.com/solana-labs/solana/blob/7b5835ddf/runtime/src/bank.rs#L3471-L3478

Reply:
I'm not really sure what you're asking. Which logic are you worried that we're relying on in other places?

Reply:
I mean here:
https://github.com/solana-labs/solana/blob/7b5835ddf/runtime/src/bank.rs#L3471-L3478
it is zeroing out the data in addition to `lamports = 0`.

Are there any places in the code that distinguish between these two accounts:
* an account with `lamports == 0` and `data == [0; size]`
* an account with `lamports == 0` but non zero data;

Reply:
Oh I see, good point. Another place would be eager rent collection which just sets lamports to 0.  Can't think of any others